capstone/src/main/java/com/google/sps/servlets/GroupPostDataServlet.java

```diff
@@ -45,43 +47,94 @@
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     // Receives submitted post 
-    String authorName = "Jane Doe";
+    Long groupId = Long.parseLong(request.getParameter("groupId"));
+    Entity userEntity = getUserFromId(response, userId, datastore);
+    String authorName = userEntity.getProperty("firstName") + " " + userEntity.getProperty("lastName");
     String postText = request.getParameter("post-input");
     String challengeName = "Challenge Name";
     String img = getUploadedFileUrl(request, "image");
     HashSet<String> likes = new HashSet<>();
     ArrayList<Comment> comments = new ArrayList<>();
 
     // Creates entity with submitted data and add to database
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     Post post = new Post(0, authorName, postText, comments, challengeName, System.currentTimeMillis(), img, likes);
-    datastore.put(post.toEntity());
+    Entity postEntity = post.toEntity();
+    datastore.put(postEntity);
+
+    // Add postsId to group
+    addPostToGroup(response, groupId, datastore, postEntity);
 
     // Redirect back to the HTML page.
-    response.sendRedirect("/group.html");
+    response.sendRedirect("/group.html?groupId="+groupId);
+  }
+
+  private void addPostToGroup(HttpServletResponse response, Long groupId, DatastoreService datastore, Entity postEntity) throws IOException {
+    Entity groupEntity = getGroupFromId(response, groupId, datastore);
+    ArrayList<Long> postIds = 
+        (ArrayList<Long>) groupEntity.getProperty("posts");
+    if (postIds == null) {
+      postIds = new ArrayList<Long>();
+    }
+    Post storedPost = Post.fromEntity(postEntity);
+    postIds.add(storedPost.getPostId());
+    groupEntity.setProperty("posts", postIds);
+    datastore.put(groupEntity);
+  }
+
+  private Entity getGroupFromId(
+    HttpServletResponse response, long groupId, DatastoreService datastore) throws IOException {
+    try {
+      return datastore.get(KeyFactory.createKey("Group", groupId));
+    } catch (EntityNotFoundException e) {
+      ErrorHandler.sendError(response, "Group does not exist.");
+      return null;
+    }
+  }
+
+  private Entity getUserFromId(
+    HttpServletResponse response, String userId, DatastoreService datastore) throws IOException {
+    try {
+      return datastore.get(KeyFactory.createKey("User", userId));
+    } catch (EntityNotFoundException e) {
+      ErrorHandler.sendError(response, "User does not exist.");
+      return null;
+    }
+  }
+
+  private Entity getPostFromId(
+    HttpServletResponse response, Long postId, DatastoreService datastore) throws IOException {
+    try {
+      return datastore.get(KeyFactory.createKey("Post", postId));
+    } catch (EntityNotFoundException e) {
+      ErrorHandler.sendError(response, "Post does not exist.");
+      return null;
+    }
```

Comment:
You can use the ServletHelper.getEntityFromId method instead of needing to create these methods for each entity